iwyu_tool.py

```diff
@@ -197,41 +225,31 @@ def slice_compilation_db(compilation_db, selection):
 
     return new_db
 
+def execute(invocations, verbose, formatter, jobs):
+    if jobs == 1:
+        for invocation in invocations:
+            print(formatter(invocation.run(verbose).get_output()))
+        return
 
-def _run_iwyu(invocation, verbose):
-    """ Start the process described by invocation.
-
-    Only necessary because multiprocessing can't pickle instance methods.
-    """
-    return invocation.run(verbose)
-
+    pending = []
+    while invocations or pending:
+        # get complete iwyu processes
+        complete = [proc for proc in pending if proc.poll() is not None]
 
-def execute(invocations, verbose, formatter, jobs):
-    """ Execute all invocations with at most `jobs` in parallel. """
-    try:
-        pool = multiprocessing.Pool(jobs)
-        # No actual results in `results`, it's only used for exception handling.
-        # Details here: https://stackoverflow.com/a/28660669.
-        results = []
+        for proc in complete:
+            # handle iwyu output
+            pending.remove(proc)
+            print(formatter(proc.get_output()))
+            proc.close()
 
-        for invocation in invocations:
-            results.append(
-                pool.apply_async(
-                    _run_iwyu, (invocation, verbose),
-                    callback=lambda x: print(formatter(x))))
-        pool.close()
-        pool.join()
-        for r in results:
-            r.get()
-    except OSError as why:
-        print('ERROR: Failed to launch include-what-you-use: %s' % why)
-        return 1
+        n = jobs - len(pending)
+        pending.extend(e.run(verbose) for e in invocations[:n])
+        invocations = invocations[n:]
 
-    return 0
+        time.sleep(0) # yield cpu
```

Comment:
This doesn't work on my Ubuntu system at least. Needs to be a longer sleep to actually yield.

Reply:
You're giving me mixed signals here ;)

Reply:
About what? I've (tried to) note several times that `sleep(0)` does not yield. When I run this script with `top` in another terminal, Python is competing with `include-what-you-use` for CPU. I wish there was a way in Python to just yield, but it appears to need a non-zero wait.

Reply:
It occurred to me that maybe you could use `subprocess.Popen.wait` instead of `poll` and provide a zero timeout. If that doesn't work, increase the timeout slightly. Waiting for processes is usually blocking.